File: 0213-house-robber-ii/0213-house-robber-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    // Dynamic Programming bottom up approach - tabulation
    const n = nums.length;
    if (n === 1) return nums[0];
    return Math.max(helper1(), helper2());
    
    function helper1() {
        // Do not try the last house
        const dp = new Array(n-1).fill(0);        
        dp[0] = nums[0];
        dp[1] = Math.max(dp[0], nums[1]);
        for (let i=2; i<n-1; i++) {
            dp[i] = Math.max(dp[i-2] + nums[i], dp[i-1]);
        }
        return dp[n-2];
    }
    
    function helper2() {
        //skip first house
        const dp = new Array(n-1).fill(0);        
        dp[1] = nums[1];
        dp[2] = Math.max(dp[1], nums[2]);
        for (let i=3; i<n; i++) {
            dp[i] = Math.max(dp[i-2] + nums[i], dp[i-1]);
        }
        return dp[n-1];
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
      # Helper function 1
      def rob1(nums, i, memo):
        if i in memo:
          return memo[i]
        if i >= len(nums) - 1:
          return 0

        include = nums[i] + rob1(nums, i+2, memo)
        exclude = rob1(nums, i+1, memo)
        memo[i] = max(include, exclude)

        return memo[i]

      # Helper function 2
      def rob2(nums, j, memo):
        if j in memo:
          return memo[j]
        if j >= len(nums):
          return 0

        include = nums[j] + rob2(nums, j+2, memo)
        exclude = rob2(nums, j+1, memo)
        memo[j] = max(include, exclude)

        return memo[j]

      
      return max(nums[0], rob1(nums, 0, {}), rob2(nums, 1, {}))  
        
        
        
        
        
        